domain/src/main/java/org/oppia/domain/topic/StoryProgressController.kt

```diff
@@ -32,27 +42,229 @@ private val FRACTIONS_COMPLETED_CHAPTERS = listOf(FRACTIONS_EXPLORATION_ID_0)
+    SUCCESS,
+    EXPLORATION_ALREADY_COMPLETED,
+    EXPLORATION_NOT_FOUND,
+    STORY_NOT_FOUND,
+    TOPIC_NOT_FOUND,
+    STORY_PROGRESS_NOT_FOUND
+  }
+
   private val trackedStoriesProgress: Map<String, TrackedStoryProgress> by lazy { createInitialStoryProgressState() }
 
+  private val cacheStoreMap = mutableMapOf<ProfileId, PersistentCacheStore<TopicProgressDatabase>>()
+
+  private fun retrieveCacheStore(profileId: ProfileId): PersistentCacheStore<TopicProgressDatabase> {
+    return if (profileId in cacheStoreMap) {
+      cacheStoreMap[profileId]!!
+    } else {
+      val cacheStore =
+        cacheStoreFactory.createPerProfile(CACHE_NAME, TopicProgressDatabase.getDefaultInstance(), profileId)
+      cacheStoreMap[profileId] = cacheStore
+      cacheStore
+    }
+  }
+
+  fun retrieveStoryProgress(storyId: String): StoryProgress {
+    return createStoryProgressSnapshot(storyId)
+  }
+
   /**
-   * Records the specified chapter completed within the context of the specified story. Returns a [LiveData] that
+   * Records the specified chapter completed within the context of the specified exploration, story, topic. Returns a [LiveData] that
    * provides exactly one [AsyncResult] to indicate whether this operation has succeeded. This method will never return
    * a pending result.
    */
-  fun recordCompletedChapter(storyId: String, explorationId: String): LiveData<AsyncResult<Nothing?>> {
-    return try {
-      trackCompletedChapter(storyId, explorationId)
-      MutableLiveData(AsyncResult.success(null))
-    } catch (e: Exception) {
-      MutableLiveData(AsyncResult.failed(e))
+  fun recordCompletedChapter(
+    internalProfileId: Int,
+    topicId: String,
+    storyId: String,
+    explorationId: String
+  ): LiveData<AsyncResult<Any?>> {
+    val profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+    val deferred = retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(updateInMemoryCache = true) {
+      // Replace this with checking if the id pair already exists.
+      val chapterPlayState = ChapterPlayState.COMPLETED
+      val storyProgressNew = StoryProgress.newBuilder().putChapterProgress(explorationId, chapterPlayState).build()
+      val topicProgress = TopicProgress.newBuilder().putStoryProgress(storyId, storyProgressNew).build()
+      val topicDatabaseBuilder = it.toBuilder().putTopicProgress(topicId, topicProgress)
+      Pair(topicDatabaseBuilder.build(), StoryProgressActionStatus.SUCCESS)
+    }
+    return dataProviders.convertToLiveData(
+      dataProviders.createInMemoryDataProviderAsync(ADD_STORY_PROGRESS_TRANSFORMED_PROVIDER_ID) {
+        return@createInMemoryDataProviderAsync getDeferredResult(
+          profileId,
+          explorationId,
+          storyId,
+          topicId,
+          deferred
+        )
+      })
+  }
+
+  fun getTopicProgress(internalProfileId: Int, topicId: String): LiveData<AsyncResult<TopicProgress>> {
```

Comment:
Since this part of the public API of this class, please add a KDoc (or mark it as ``internal``): https://developer.android.com/kotlin/style-guide#usage.